recursion/circuit-v2/src/fri.rs

```diff
@@ -0,0 +1,733 @@
+
+        run_test_recursion(builder.operations);
+    }
+
+    #[test]
+    fn test_verify_two_adic_pcs() {
+        let mut rng = StdRng::seed_from_u64(0xDEADBEEF);
+        let log_degrees = &[19, 19];
+        let perm = inner_perm();
+        let fri_config = inner_fri_config();
+        let hash = InnerHash::new(perm.clone());
+        let compress = InnerCompress::new(perm.clone());
+        let val_mmcs = InnerValMmcs::new(hash, compress);
+        let dft = InnerDft {};
+        let pcs: InnerPcs = InnerPcs::new(
+            log_degrees.iter().copied().max().unwrap(),
+            dft,
+            val_mmcs,
+            fri_config,
+        );
+
+        // Generate proof.
+        let domains_and_polys = log_degrees
+            .iter()
+            .map(|&d| {
+                (
+                    <InnerPcs as Pcs<InnerChallenge, InnerChallenger>>::natural_domain_for_degree(
+                        &pcs,
+                        1 << d,
+                    ),
+                    RowMajorMatrix::<InnerVal>::rand(&mut rng, 1 << d, 100),
+                )
+            })
+            .collect::<Vec<_>>();
+        let (commit, data) = <InnerPcs as Pcs<InnerChallenge, InnerChallenger>>::commit(
+            &pcs,
+            domains_and_polys.clone(),
+        );
+        let mut challenger = InnerChallenger::new(perm.clone());
+        challenger.observe(commit);
+        let zeta = challenger.sample_ext_element::<InnerChallenge>();
+        let points = domains_and_polys
+            .iter()
+            .map(|_| vec![zeta])
+            .collect::<Vec<_>>();
+        let (opening, proof) = pcs.open(vec![(&data, points)], &mut challenger);
+
+        // Verify proof.
+        let mut challenger = InnerChallenger::new(perm.clone());
+        challenger.observe(commit);
+        let x1 = challenger.sample_ext_element::<InnerChallenge>();
+        let os = domains_and_polys
+            .iter()
+            .zip(&opening[0])
+            .map(|((domain, _), mat_openings)| (*domain, vec![(zeta, mat_openings[0].clone())]))
+            .collect::<Vec<_>>();
+        pcs.verify(vec![(commit, os.clone())], &proof, &mut challenger)
+            .unwrap();
+
+        // Define circuit.
+        let mut builder = Builder::<InnerConfig>::default();
+        let config = inner_fri_config();
+        let proof = const_two_adic_pcs_proof(&mut builder, proof);
+        let (commit, rounds) = const_two_adic_pcs_rounds(&mut builder, commit.into(), os);
+        let mut challenger = DuplexChallengerVariable::new(&mut builder);
+        challenger.observe_commitment(&mut builder, commit);
+        let x2 = challenger.sample_ext(&mut builder);
+        let x1: Ext<_, _> = builder.constant(x1);
+        builder.assert_ext_eq(x1, x2);
+        verify_two_adic_pcs(&mut builder, &config, &proof, &mut challenger, rounds);
+
+        run_test_recursion(builder.operations);
+        // let mut backend = ConstraintCompiler::<InnerConfig>::default();
```

Comment:
nit: do we want to comment this in?

Reply:
It doesn't work because it calls an unsupported instruction `CircuitV2Poseidon2PermuteBabyBearWide`, and more generally my runtime, compiler, etc. currently do not support Bn254.